File: backend/app/services/post_service.py
from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.models.user import User
from app.repositories import post_repository
from app.schemas.post import PostCreate, PostUpdate


def list_posts(db: Session, page: int, limit: int, search: str | None):
    return post_repository.list_posts(db, page, limit, search)


def list_my_posts(db: Session, user: User):
    return post_repository.list_posts_by_author(db, user.id)


def get_post(db: Session, post_id: int):
    post = post_repository.get_post(db, post_id)
    if not post:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Post not found")
    return post


def create_post(db: Session, payload: PostCreate, user: User):
    return post_repository.create_post(db, payload, user.id)


def update_post(db: Session, post_id: int, payload: PostUpdate, user: User):
    post = get_post(db, post_id)
    if post.author_id != user.id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You can only edit your own posts")
    return post_repository.update_post(db, post, payload)


def delete_post(db: Session, post_id: int, user: User) -> None:
    post = get_post(db, post_id)
    if post.author_id != user.id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="You can only delete your own posts")
    post_repository.delete_post(db, post)

File: frontend/src/services/postService.js
import { api } from '../api/client.js'

export const postService = {
  async list({ page = 1, limit = 10, search = '' } = {}) {
    const params = new URLSearchParams({ page, limit })
    if (search) params.set('search', search)
    const { data } = await api.get(`/posts?${params.toString()}`)
    return data
  },
  async get(id) {
    const { data } = await api.get(`/posts/${id}`)
    return data
  },
  async create(payload) {
    const { data } = await api.post('/posts', payload)
    return data
  },
  async update(id, payload) {
    const { data } = await api.put(`/posts/${id}`, payload)
    return data
  },
  async remove(id) {
    await api.delete(`/posts/${id}`)
  }
}
